run.go

```diff
@@ -0,0 +1,248 @@
+		}
+
+		start := time.Now()
+		if err := p.runIndividually(ctx, conn, direction.ToBool(), m); err != nil {
+			current.Error = err
+			current.Duration = time.Since(start)
+			return nil, &PartialError{
+				Results: results,
+				Failed:  current,
+				Err:     err,
+			}
+		}
+
+		current.Duration = time.Since(start)
+		results = append(results, current)
+	}
+	return results, nil
+}
+
+// runIndividually runs an individual migration, opening a new transaction if the migration is safe
+// to run in a transaction. Otherwise, it runs the migration outside of a transaction with the
+// supplied connection.
+func (p *Provider) runIndividually(
+	ctx context.Context,
+	conn *sql.Conn,
+	direction bool,
+	m *migration.Migration,
+) error {
+	if m.UseTx() {
+		// Run the migration in a transaction.
+		return p.beginTx(ctx, conn, func(tx *sql.Tx) error {
+			if err := m.Run(ctx, tx, direction); err != nil {
+				return err
+			}
+			if p.opt.NoVersioning {
+				return nil
+			}
+			return p.store.InsertOrDelete(ctx, tx, direction, m.Version)
+		})
+	}
+	// Run the migration outside of a transaction.
+	switch {
+	case m.IsGo():
+		// Note, we're using *sql.DB instead of *sql.Conn because it's the contract of the
+		// GoMigrationNoTx function. This may be a deadlock scenario if the caller sets max open
+		// connections to 1. See the comment in runMigrations for more details.
+		if err := m.RunNoTx(ctx, p.db, direction); err != nil {
+			return err
+		}
+	case m.IsSQL():
+		if err := m.RunConn(ctx, conn, direction); err != nil {
+			return err
+		}
+	}
+	if p.opt.NoVersioning {
+		return nil
+	}
+	return p.store.InsertOrDeleteConn(ctx, conn, direction, m.Version)
+}
+
+// beginTx begins a transaction and runs the given function. If the function returns an error, the
+// transaction is rolled back. Otherwise, the transaction is committed.
+//
+// If the provider is configured to use versioning, this function also inserts or deletes the
+// migration version.
+func (p *Provider) beginTx(
+	ctx context.Context,
+	conn *sql.Conn,
+	fn func(tx *sql.Tx) error,
+) (retErr error) {
+	tx, err := conn.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if retErr != nil {
+			retErr = errors.Join(retErr, tx.Rollback())
+		}
+	}()
+	if err := fn(tx); err != nil {
+		return err
+	}
+	return tx.Commit()
+}
+
+func (p *Provider) initialize(ctx context.Context) (*sql.Conn, func() error, error) {
+	p.mu.Lock()
+
+	conn, err := p.db.Conn(ctx)
+	if err != nil {
+		p.mu.Unlock()
+		return nil, nil, err
+	}
+	var (
+		// cleanup is a function that cleans up the connection, and optionally, the session lock.
+		cleanup func() error
+	)
+	switch p.opt.LockMode {
+	case LockModeAdvisorySession:
+		if err := p.store.LockSession(ctx, conn); err != nil {
+			p.mu.Unlock()
+			return nil, nil, err
+		}
+		cleanup = func() error {
+			defer p.mu.Unlock()
+			return errors.Join(p.store.UnlockSession(ctx, conn), conn.Close())
+		}
```

Comment:
You're spot on, that's exactly how it'd work. app 2 would acquire the lock, and if app 1 is restarted (for any reason) it would fail to come up because the lock is held by the (not-yet-ready) app 2.

By status, do you mean comparing against the last applied migration? I think this would only work for sequential migrations, but we'd also have to account for out-of-order migrations.

A crude way would be to list all db versions and compared them against the migration files. If there's 100% parity then we can exit early. I believe this would work for both sequential and out-of-order migrations.

There's an edge case here, if app 2 has two migrations to apply, a fast one and a long-running one, then app 1 could see a view of the world that is different.

E.g., we have 1,2,3 applied and then app 2 starts running with migrations 4, 5. 

- migration 4 gets applied immediately.
- migration 5 is long-running
- app 1 is restarted, it sees 1,2,3,4 in the goose db table.. but it only knows about 1,2,3
- should app 1 exit with no error here?

Reply:
Thanks again for the detailed report, I went ahead and implemented this in https://github.com/pressly/goose/pull/751 and it's natively baked into the `*goose.Provider`, exposing both a `.HasPending()` method and also a pre-check in the `.Up` set of commands.